File: app/src/test/java/com/coinninja/coinkeeper/util/CurrencyPreferenceTest.java
package com.coinninja.coinkeeper.util;

import android.content.Intent;

import com.coinninja.coinkeeper.factory.CurrencyFactory;
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil;
import com.coinninja.coinkeeper.util.android.PreferencesUtil;
import com.coinninja.coinkeeper.util.currency.BTCCurrency;
import com.coinninja.coinkeeper.util.currency.USDCurrency;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static com.coinninja.matchers.IntentMatcher.equalTo;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class CurrencyPreferenceTest {

    @InjectMocks
    CurrencyPreference currencyPreference;
    @Mock
    private CurrencyFactory currencyFactory;
    @Mock
    private PreferencesUtil preferencesUtil;
    @Mock
    private LocalBroadCastUtil localBroadCastUtil;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @After
    public void tearDown() {
        localBroadCastUtil = null;
        currencyFactory = null;
        preferencesUtil = null;
        currencyPreference = null;
    }

    @Test
    public void sets_base_currency_preference() {
        BTCCurrency primary = new BTCCurrency();
        USDCurrency secondary;
        secondary = new USDCurrency();

        currencyPreference.setCurrencies(primary, secondary);

        verify(preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, primary.getSymbol());
        verify(preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, secondary.getSymbol());
    }

    @Test
    public void provides_access_to_currency_preference() {
        when(preferencesUtil.getString(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, BTCCurrency.SYMBOL))
                .thenReturn(BTCCurrency.SYMBOL);
        when(preferencesUtil.getString(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, USDCurrency.SYMBOL))
                .thenReturn(USDCurrency.SYMBOL);

        when(currencyFactory.fromSymbol(BTCCurrency.SYMBOL)).thenReturn(new BTCCurrency());
        when(currencyFactory.fromSymbol(USDCurrency.SYMBOL)).thenReturn(new USDCurrency());

        DefaultCurrencies defaultCurrencies = currencyPreference.getCurrenciesPreference();

        assertThat(defaultCurrencies.getPrimaryCurrency().getSymbol(), equalTo(BTCCurrency.SYMBOL));
        assertThat(defaultCurrencies.getSecondaryCurrency().getSymbol(), equalTo(USDCurrency.SYMBOL));
    }

    @Test
    public void allows_toggling_of_default() {
        when(preferencesUtil.getString(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, BTCCurrency.SYMBOL))
                .thenReturn(BTCCurrency.SYMBOL);
        when(preferencesUtil.getString(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, USDCurrency.SYMBOL))
                .thenReturn(USDCurrency.SYMBOL);

        when(currencyFactory.fromSymbol(BTCCurrency.SYMBOL)).thenReturn(new BTCCurrency());
        when(currencyFactory.fromSymbol(USDCurrency.SYMBOL)).thenReturn(new USDCurrency());

        DefaultCurrencies defaultCurrencies = currencyPreference.toggleDefault();

        verify(preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, USDCurrency.SYMBOL);
        verify(preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, BTCCurrency.SYMBOL);
        assertThat(defaultCurrencies.getPrimaryCurrency().getSymbol(), equalTo(USDCurrency.SYMBOL));
        assertThat(defaultCurrencies.getSecondaryCurrency().getSymbol(), equalTo(BTCCurrency.SYMBOL));
    }

    @Test
    public void dispatches_local_notification_that_preference_changed() {
        ArgumentCaptor<Intent> argumentCaptor = ArgumentCaptor.forClass(Intent.class);
        DefaultCurrencies defaultCurrencies = new DefaultCurrencies(new USDCurrency(), new BTCCurrency());
        Intent expected = new Intent(DropbitIntents.ACTION_CURRENCY_PREFERENCE_CHANGED);
        expected.putExtra(DropbitIntents.EXTRA_PREFERENCE, defaultCurrencies);
        when(preferencesUtil.getString(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, BTCCurrency.SYMBOL)).thenReturn(USDCurrency.SYMBOL);
        when(preferencesUtil.getString(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, USDCurrency.SYMBOL)).thenReturn(BTCCurrency.SYMBOL);
        when(currencyFactory.fromSymbol(BTCCurrency.SYMBOL)).thenReturn(new BTCCurrency());
        when(currencyFactory.fromSymbol(USDCurrency.SYMBOL)).thenReturn(new USDCurrency());

        currencyPreference.setCurrencies(defaultCurrencies.getPrimaryCurrency(), defaultCurrencies.getSecondaryCurrency());

        verify(localBroadCastUtil).sendBroadcast(argumentCaptor.capture());
        Intent intent = argumentCaptor.getValue();
        assertThat(intent, equalTo(expected));
    }

}

File: app/src/test/java/com/coinninja/coinkeeper/util/CurrencyPreferenceTest.kt
package com.coinninja.coinkeeper.util

import android.content.Intent
import androidx.test.ext.junit.runners.AndroidJUnit4
import app.dropbit.commons.currency.BTCCurrency
import app.dropbit.commons.currency.USDCurrency
import com.coinninja.matchers.IntentMatcher
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.verify

@RunWith(AndroidJUnit4::class)
class CurrencyPreferenceTest {
    private fun createCurrencyPreference(): CurrencyPreference = CurrencyPreference(mock(), mock(), mock())

    @Test
    fun sets_base_currency_preference() {
        val primary = BTCCurrency()
        val secondary: USDCurrency
        secondary = USDCurrency()
        val currencyPreference = createCurrencyPreference()

        currencyPreference.setCurrencies(primary, secondary)

        verify(currencyPreference.preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, primary.symbol)
        verify(currencyPreference.preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, secondary.symbol)
    }

    @Test
    fun provides_access_to_currency_preference() {
        val currencyPreference = createCurrencyPreference()
        whenever(currencyPreference.preferencesUtil.getString(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, USDCurrency.symbol))
                .thenReturn(USDCurrency.symbol)
        whenever(currencyPreference.preferencesUtil.getString(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, BTCCurrency.symbol))
                .thenReturn(BTCCurrency.symbol)

        whenever(currencyPreference.currencyFactory.fromSymbol(BTCCurrency.symbol)).thenReturn(BTCCurrency())
        whenever(currencyPreference.currencyFactory.fromSymbol(USDCurrency.symbol)).thenReturn(USDCurrency())

        val defaultCurrencies = currencyPreference.currenciesPreference

        assertThat(defaultCurrencies.primaryCurrency.symbol, equalTo(USDCurrency.symbol))
        assertThat(defaultCurrencies.secondaryCurrency.symbol, equalTo(BTCCurrency.symbol))
    }

    @Test
    fun allows_toggling_of_default() {
        val currencyPreference = createCurrencyPreference()
        whenever(currencyPreference.preferencesUtil.getString(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, USDCurrency.symbol))
                .thenReturn(BTCCurrency.symbol)
        whenever(currencyPreference.preferencesUtil.getString(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, BTCCurrency.symbol))
                .thenReturn(USDCurrency.symbol)

        whenever(currencyPreference.currencyFactory.fromSymbol(BTCCurrency.symbol)).thenReturn(BTCCurrency())
        whenever(currencyPreference.currencyFactory.fromSymbol(USDCurrency.symbol)).thenReturn(USDCurrency())

        val defaultCurrencies = currencyPreference.toggleDefault()

        verify(currencyPreference.preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, USDCurrency.symbol)
        verify(currencyPreference.preferencesUtil).savePreference(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, BTCCurrency.symbol)
        assertThat(defaultCurrencies.primaryCurrency.symbol, equalTo(USDCurrency.symbol))
        assertThat(defaultCurrencies.secondaryCurrency.symbol, equalTo(BTCCurrency.symbol))
    }

    @Test
    fun dispatches_local_notification_that_preference_changed() {
        val currencyPreference = createCurrencyPreference()
        val argumentCaptor = ArgumentCaptor.forClass(Intent::class.java)
        val defaultCurrencies = DefaultCurrencies(USDCurrency(), BTCCurrency())
        val expected = Intent(DropbitIntents.ACTION_CURRENCY_PREFERENCE_CHANGED)
        expected.putExtra(DropbitIntents.EXTRA_PREFERENCE, defaultCurrencies)
        whenever(currencyPreference.preferencesUtil.getString(CurrencyPreference.PREFERENCE_PRIMARY_CURRENCY, USDCurrency.symbol))
                .thenReturn(USDCurrency.symbol)
        whenever(currencyPreference.preferencesUtil.getString(CurrencyPreference.PREFERENCE_SECONDARY_CURRENCY, BTCCurrency.symbol))
                .thenReturn(BTCCurrency.symbol)
        whenever(currencyPreference.currencyFactory.fromSymbol(BTCCurrency.symbol)).thenReturn(BTCCurrency())
        whenever(currencyPreference.currencyFactory.fromSymbol(USDCurrency.symbol)).thenReturn(USDCurrency())

        currencyPreference.setCurrencies(defaultCurrencies.primaryCurrency, defaultCurrencies.secondaryCurrency)

        verify(currencyPreference.localBroadCastUtil).sendBroadcast(argumentCaptor.capture())
        val intent = argumentCaptor.value
        assertThat(intent, IntentMatcher.equalTo(expected))
    }

}